src/dist/failure_detector/failure_detector.cpp

```diff
@@ -340,37 +339,42 @@ void failure_detector::end_ping(::dsn::error_code err, const beacon_ack& ack, vo
 
     master_map::iterator itr = _masters.find(node);
 
-    if ( itr == _masters.end() )
+    if (itr == _masters.end())
     {
         dwarn("received beacon ack without corresponding master, ignore it, "
-              "remote_master[%s], local_worker[%s]",
-              node.to_string(), primary_address().to_string());
-
-        return;
+            "remote_master[%s], local_worker[%s]",
+            node.to_string(), primary_address().to_string());
+        return false;
     }
 
     master_record& record = itr->second;
     if (!ack.allowed)
     {
         dwarn("worker rejected, stop sending beacon message, "
-              "remote_master[%s], local_worker[%s]",
-              node.to_string(), primary_address().to_string());
+            "remote_master[%s], local_worker[%s]",
+            node.to_string(), primary_address().to_string());
         record.rejected = true;
-        return;
+        return false;
     }
 
-    if (is_time_greater_than(beacon_send_time, record.last_send_time_for_beacon_with_ack))
+    if (!is_time_greater_than(beacon_send_time, record.last_send_time_for_beacon_with_ack))
```

Comment:
I still wonder that version for the FD beacon messages is necessary as a newer timeout message switches the master, but an older OK-response is handled.
For example, A worker send b1 and b2 to master1, b1 times out, but b2 OK. And end ping of b1 and b2 are handled nearly at the same time. If b1 is handled first, then b2 should be ignored. But currently it is not.